src/flat.rs

```diff
@@ -0,0 +1,1139 @@
+    /// dimension overflows `usize` with its stride we also consider this aliasing.
+    pub fn has_aliased_samples(&self) -> bool {
+        let grouped = self.increasing_stride_dims();
+        let (min_dim, mid_dim, max_dim) = (grouped[0], grouped[1], grouped[2]);
+
+        let min_size = match min_dim.checked_len() {
+            None => return true,
+            Some(size) => size,
+        };
+
+        let mid_size = match mid_dim.checked_len() {
+            None => return true,
+            Some(size) => size,
+        };
+
+        let _max_size = match max_dim.checked_len() {
+            None => return true,
+            Some(_) => (), // Only want to know this didn't overflow.
+        };
+
+        // Each higher dimension must walk over all of one lower dimension.
+        min_size > mid_dim.stride() || mid_size > max_dim.stride()
+    }
+
+    /// Check if a buffer fulfills the requirements of a normal form.
+    ///
+    /// Certain conversions have preconditions on the structure of the sample buffer that are not
+    /// captured (by design) by the type system. These are then checked before the conversion. Such
+    /// checks can all be done in constant time and will not inspect the buffer content. You can
+    /// perform these checks yourself when the conversion is not required at this moment but maybe
+    /// still performed later.
+    pub fn is_normal(&self, form: NormalForm) -> bool {
+        if self.has_aliased_samples() {
+            return false;
+        }
+
+        if form >= NormalForm::PixelPacked && self.channel_stride != 1 {
+            return false;
+        }
+
+        if form >= NormalForm::ImagePacked {
+            // has aliased already checked for overflows.
+            let grouped = self.increasing_stride_dims();
+            let (min_dim, mid_dim, max_dim) = (grouped[0], grouped[1], grouped[2]);
+
+            if 1 != min_dim.stride() {
+                return false;
+            }
+
+            if min_dim.len() != mid_dim.stride() {
+                return false;
+            }
+
+            if  mid_dim.len() != max_dim.stride() {
+                return false;
+            }
+        }
+
+        if form >= NormalForm::RowMajorPacked {
+            if self.width_stride != self.channels as usize {
+                return false;
+            }
+
+            if self.width as usize*self.width_stride != self.height_stride {
+                return false;
+            }
+        }
+
+        if form >= NormalForm::ColumnMajorPacked {
+            if self.height_stride != self.channels as usize {
+                return false;
+            }
+            
+            if self.height as usize*self.height_stride != self.width_stride {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /// Check that the pixel and the channel index are in bounds.
+    pub fn in_bounds(&self, x: u32, y: u32, channel: u8) -> bool {
```

Comment:
One concern of myself, which I am not yet sure how to resolve, is how the order here and in `index` is inconsistent with `strides_hwc` and `extents`. We want to stay consistent with both `Image::get(x, y)` and the order of potentially row-major strides (`hwc` or `cwh`). This suggest switching to `fn in_bounds(&self, channel: u8, x: u32, y: u32)`.